### PR TITLE
Rewrite test cases for instant search.

### PR DESCRIPTION
**Pull request by brendanator**

### CODE REVIEW

**Comment by brendanator** [comment]
This is a benchmark review for experiment `review_of_100_reviews_20240409`.
Run ID: `review_of_100_reviews_20240409/benchmark_2024-04-09T10-24-34_v1-16-0-95-g65fefaeef-dirty`.

This pull request was cloned from `https://github.com/suttacentral/suttacentral/pull/3122`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

File: Makefile [hunk 1]
```diff
@@ -60,7 +60,7 @@ test:
 	@make test-server
 
 test-server:
-	@docker exec -t sc-flask pytest server/
```
Comment (brendanator) [inline Makefile:63]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Adding `-s` flag to pytest may expose sensitive information in logs.

The `-s` flag disables all capturing. Ensure that no sensitive data is printed to the console as it could be exposed in CI/CD logs.

File: server/server/search/tests/test_instant_search.py [hunk 2]
```diff
@@ -120,4 +124,52 @@ def test_normalize_string():
     assert normalize_string('rupa') == 'rupa'
     assert normalize_string('Rūpā') == 'Rupa'
     assert normalize_string('samma') == 'samma'
-    assert normalize_string('sāmma') == 'samma'
\ No newline at end of file
+    assert normalize_string('sāmma') == 'samma'
+
+
+def instant_search(query, languages):
```
Comment (brendanator) [inline server/server/search/tests/test_instant_search.py:130]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Consider parameterizing `instant_search` tests for better maintainability.

Using pytest's parameterize feature could help reduce redundancy and make it easier to add new test cases in the future.

Comment (brendanator) [inline server/server/search/tests/test_instant_search.py:130]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (testing):** The `instant_search` function lacks assertions for specific expected results.

While checking for a non-zero total is a good start, more specific assertions (e.g., checking for expected items in the response) would ensure the search functionality works as intended.

Comment (brendanator) [inline server/server/search/tests/test_instant_search.py:130]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Missing test cases for error conditions and edge cases.

Consider adding tests for invalid queries, unsupported languages, and queries that should return zero results to ensure robust error handling and edge case coverage.

```suggestion
def test_invalid_query_search():
    instant_search('', ['en', 'pli'])  # Empty query
    instant_search(' ', ['en', 'pli'])  # Query with only spaces

def test_unsupported_language_search():
    instant_search('dhamma', ['xx'])  # Unsupported language code

def test_zero_results_search():
    instant_search('nonexistentterm123', ['en', 'pli'])  # Query unlikely to return results
```

Comment (brendanator) [inline server/server/search/tests/test_instant_search.py:130]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** The test does not validate the structure of the JSON response.

Validating the expected structure of the JSON response, including key names and types of values, would help catch serialization and deserialization issues.

Comment (brendanator) [inline server/server/search/tests/test_instant_search.py:130]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** No tests for unauthorized access.

It's important to ensure that the search API handles unauthorized access correctly, either by returning an appropriate error code or by not exposing sensitive information.

```suggestion
def test_unauthorized_access():
    response = client.post('/search/instant?query=test&language=en&limit=10&offset=0&matchpartial=false')
    assert response.status_code == 401 or response.status_code == 403
    assert 'error' in response.get_json()
```

**Review by brendanator — COMMENTED** [review]
<a href="https://sourcery.ai">Sourcery logo</a>  Sourcery Review:

Hey @brendanator - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟢 **General issues**: all looks good
- 🟡 **Security**: 1 issue found
- 🟡 **Testing**: 5 issues found
- 🟢 **Complexity**: all looks good
- 🟢 **Docstrings**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/5c8bea2b-6c6c-4221-bd9b-0c11556933a8/projects/p/2dbf7697-3bb3-4f63-b184-3a8ebf7ac23c/r/232470ce-4e4d-4558-b336-2251f4be70c0?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

File: Makefile (hunk 1 above)
Comment (SourceryAI) [reply Makefile:63]: Is this comment correct?

Comment (SourceryAI) [reply Makefile:63]: Is this comment helpful?

Comment (SourceryAI) [reply Makefile:63]: Is the comment type correct?

Comment (SourceryAI) [reply Makefile:63]: Is the comment area correct?

File: server/server/search/tests/test_instant_search.py (hunk 2 above)
Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment helpful?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment type correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment area correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment helpful?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment type correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment area correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment helpful?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment type correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is the comment area correct?

Comment (SourceryAI) [reply server/server/search/tests/test_instant_search.py:130]: Is this comment correct?
